File: slp/backend-dotnet/Features/Source/SourceDTO.cs
using System;

namespace backend_dotnet.Features.Source;

public class SourceDto
{
    public int Id { get; set; }
    public int UserId { get; set; }
    public string Type { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public string? Url { get; set; }
    public string? RawText { get; set; }

    /// <summary>
    /// TipTap JSON (stored as raw string so frontend can parse directly).
    /// Null if the source has no rich-text content.
    /// </summary>
    public string? ContentJson { get; set; }

    public string? FilePath { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public string? Metadata { get; set; }
}

public class SourceListDto
{
    public int Id { get; set; }
    public string Type { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public string? Url { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
}

public class UploadSourceRequest
{
    public required IFormFile File { get; set; }

    // FIX: Title is optional — parser or filename is used as fallback
    public string? Title { get; set; }
}

public class CreateNoteSourceRequest
{
    public string Title { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
}

public class SourceQueryParams
{
    public int Page { get; set; } = 1;
    public int PageSize { get; set; } = 20;

    /// <summary>Partial, case-insensitive match on Title.</summary>
    public string? Search { get; set; }

    /// <summary>Exact match on Type (pdf | link | text). Null = all.</summary>
    public string? Type { get; set; }
}

File: slp/backend-nestjs/src/modules/source/source.dto.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { IsOptional, IsString, IsInt, Min, Max } from 'class-validator';

export class SourceQueryParams {
  @ApiPropertyOptional({ default: 1 })
  @IsOptional()
  @IsInt()
  @Min(1)
  @Type(() => Number)
  page?: number = 1;

  @ApiPropertyOptional({ default: 20 })
  @IsOptional()
  @IsInt()
  @Min(1)
  @Max(100)
  @Type(() => Number)
  pageSize?: number = 20;

  @ApiPropertyOptional()
  @IsOptional()
  @IsString()
  search?: string;

  @ApiPropertyOptional()
  @IsOptional()
  @IsString()
  type?: string;
}

export class SourceDto {
  id: number;
  userId: number;
  type: string;
  title: string;
  url?: string | null;
  rawText?: string | null;
  contentJson?: string | null;
  filePath?: string | null;
  createdAt: Date;
  updatedAt: Date;
  metadata?: string | null;
}

export class SourceListDto {
  id: number;
  type: string;
  title: string;
  url?: string | null;
  createdAt: Date;
  updatedAt: Date;
}

export class UploadSourceRequest {
  @ApiProperty({ type: 'string', format: 'binary' })
  file: Express.Multer.File;

  @ApiPropertyOptional()
  @IsOptional()
  @IsString()
  title?: string;
}

export class UrlSourceDto {
  @ApiProperty()
  @IsString()
  url: string;

  @ApiPropertyOptional()
  @IsOptional()
  @IsString()
  title?: string;
}

export class CreateNoteSourceRequest {
  @ApiProperty()
  @IsString()
  title: string;

  @ApiProperty()
  @IsString()
  content: string;
}

// Progress-related DTOs (minimal)
export class ProgressDto {
  sourceId: number;
  progress: number; // e.g., 0.0 – 1.0
  updatedAt: Date;
}